src/Sets/LineSegment/intersection.jl

```diff
@@ -34,24 +34,34 @@ function intersection(LS1::LineSegment, LS2::LineSegment)
     m = intersection(L1, L2)
     N = promote_type(eltype(LS1), eltype(LS2))
     if m == L1
-        # determine which segment is in both
-        p1 = max(min(LS1.p[1], LS1.q[1]), min(LS2.p[1], LS2.q[1]))
-        p2 = max(min(LS1.p[2], LS1.q[2]), min(LS2.p[2], LS2.q[2]))
-        q1 = min(max(LS1.p[1], LS1.q[1]), max(LS2.p[1], LS2.q[1]))
-        q2 = min(max(LS1.p[2], LS1.q[2]), max(LS2.p[2], LS2.q[2]))
-        if _isapprox(p1, q1) && _isapprox(p2, q2)
-            return Singleton([p1, p2])  # edges have a point in common
+        # line segments are on the same line
 
-        elseif _leq(p1, q1) && _leq(p2, q2)
-            return LineSegment([p1, p2], [q1, q2])
+        # check that the line segments intersect
+        @inbounds begin
+            # box approximation
+            l, r = extrema((LS1.p[1], LS1.q[1]))
+            b, t = extrema((LS1.p[2], LS1.q[2]))
+            # check that the other line segment has at least one point inside the box
+            if !(l <= LS2.p[1] <= r && b <= LS2.p[2] <= t) &&
+               !(l <= LS2.q[1] <= r && b <= LS2.q[2] <= t)
+                return EmptySet{N}(2)
+            end
+        end
 
+        # find the middle two points of the four end points
+        points = [LS1.p, LS1.q, LS2.p, LS2.q]
```

Comment:
One could use `SVector(LS1.p, LS1.q, LS2.p, LS2.q)` I suppose but maybe at some point one doesn't have to worry about optimization anymore 😄 as it's already much faster than the current version + the bug is fixed.

Reply:
I agree. Just to mention that my measurements were wrt. my first new version. I did not benchmark the old version, which was probably similarly efficient (but, as you say, incorrect).

Reply:
Actually even comparing it to the current  v2.14.1 one we got twice as fast.

```julia
julia> ls1 = LineSegment(Float32[-34.67, -46.06], Float32[-13.03, -49.77])
LineSegment{Float32, Vector{Float32}}(Float32[-34.67, -46.06], Float32[-13.03, -49.77])

julia> ls2 = LineSegment(Float32[-22.473152, -48.13686], Float32[-19.031397, -48.730015])

julia> @benchmark intersection($ls1, $ls2) # LazySets  v2.14.1
BenchmarkTools.Trial: 10000 samples with 9 evaluations.
 Range (min … max):  2.269 μs …   9.648 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.389 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.424 μs ± 208.256 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

      ▄█▇▃▁▁▂▁                                                 
  ▁▁▂▆█████████▆▄▃▃▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  2.27 μs         Histogram: frequency by time        3.09 μs <

 Memory estimate: 1.14 KiB, allocs estimate: 47.

julia> @benchmark new_intersection($ls1, $ls2) # with SVector
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.079 μs …  3.646 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.133 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.149 μs ± 86.237 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

          ▁█▁▁  ▁                                             
  ▂▂▂▃▇▇███████▇█▅▄▄▄▅▃▃▃▃▄▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂ ▃
  1.08 μs        Histogram: frequency by time        1.37 μs <

 Memory estimate: 560 bytes, allocs estimate: 21.
 ```